### ISSUE TITLE
"Class" loaded more than once?

### ISSUE THREAD

**Issue by stevenou** [thoughtbot/shoulda-matchers#1639]
### Description

Changes introduced in https://github.com/thoughtbot/shoulda-matchers/pull/1610 are causing "the class" to load more than once. Specifically, the change to `namespace.set(name, Class.new(parent))`.

I encountered this because [searchkick](https://github.com/ankane/searchkick) now raises `RuntimeError: Only call searchkick once per model`. I am running `searchkick` inside `def inherited(subclass)` and the changes are causing that to run again, thus raising an error.

### Reproduction Steps

Not too sure what reproduction steps are relevant. At this point I don't understand the issue well enough to even know if it should be considered a bug or not.

### Expected behavior

"The class" isn't loaded more than once during testing. This is the behavior in version <= 6.1.0.

### Actual behavior

In version 6.2.0, the code inside `def inherited(subclass)` runs more than once for the "same" class, presumably because there is now another anonymous class that inherits from it. 

### System configuration
**shoulda_matchers version**: 6.2.0
**rails version**: 7.1.3.4
**ruby version**: 3.3.0

**Comment by myupchar** [comment]
Did you figure out a solution for this? Ran into same problem while upgrading

**Comment by stevenou** [comment]
> Did you figure out a solution for this? Ran into same problem while upgrading

No, I did not figure out a solution. Currently locked to version 6.1.0...

**Comment by matsales28** [comment]
Hi, @stevenou and @myupchar 

Apologies for the delay in responding—things have been quite busy on my end.

It seems the behaviour arises from the `inherited` method. While this is expected when using Class.new(parent), you can mitigate the issue by modifying the `inherited` method in your parent class with a guard clause. Here's an example:

```ruby
def inherited(subclass)
  # Skip if the subclass is anonymous (i.e., has no name)
  return if subclass.name.nil?

  # Alternatively, prevent duplicate processing with a safeguard
  @processed_subclasses ||= Set.new
  return if @processed_subclasses.include?(subclass.object_id)

  @processed_subclasses.add(subclass.object_id)

  # Your existing inherited logic here
  super
end
```

Do you think that would work for your case?

**Comment by stevenou** [comment]
@matsales28 thanks for the suggestion. this worked for me.
